hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -747,8 +740,37 @@ public void onFailure(Throwable cause) {
                 logger.finest("Authentication of " + connection + " failed.", cause);
             }
             connection.close(null, cause);
-            connectionsInProgress.remove(toInetSocketAddress(target));
+            connectionsInProgress.remove(inetSocketAddressCache.get(target));
             future.onFailure(cause);
         }
     }
+
+    private static class InetSocketAddressCache {
+
+        private final ConcurrentMap<Address, InetSocketAddress> cache = new ConcurrentHashMap<Address, InetSocketAddress>();
+
+        private InetSocketAddress get(Address target) {
+            try {
+                InetSocketAddress resolvedAddress = cache.get(target);
+                if (resolvedAddress != null) {
+                    return resolvedAddress;
+                }
+                InetSocketAddress newResolvedAddress = new InetSocketAddress(target.getInetAddress(), target.getPort());
+                InetSocketAddress prevAddress = cache.putIfAbsent(target, newResolvedAddress);
+                if (prevAddress != null) {
+                    return prevAddress;
+                }
+                return newResolvedAddress;
+            } catch (UnknownHostException e) {
+                throw rethrow(e);
```

Comment:
do you really need rethrow here?

Reply:
UnknownHostException is a checked exception, I need to handle it or propagate. Because it wil cause API changes, I need to rethrow it as RuntimeException. 
Alternative is to returning null here, which will not work. The exception needs to be catched an handled on upper levels.

Reply:
Ok